source/common/network/filter_manager_impl.h

```diff
@@ -11,35 +11,114 @@ namespace Envoy {
  */
-class BufferSource {
+class StreamBufferSource {
 public:
-  virtual ~BufferSource() {}
+  virtual ~StreamBufferSource() {}
 
   struct StreamBuffer {
     Buffer::Instance& buffer;
     bool end_stream;
   };
+};
+
+/**
+ * Interface used to obtain read buffers.
+ */
+class ReadBufferSource : public virtual StreamBufferSource {
+public:
+  virtual ~ReadBufferSource() {}
 
   /**
    * Fetch the read buffer for the source.
    */
   virtual StreamBuffer getReadBuffer() PURE;
+};
+
+/**
+ * Interface used to obtain write buffers.
+ */
+class WriteBufferSource : public virtual StreamBufferSource {
+public:
+  virtual ~WriteBufferSource() {}
 
   /**
    * Fetch the write buffer for the source.
    */
   virtual StreamBuffer getWriteBuffer() PURE;
 };
 
+/**
+ * Interface used to obtain read/write buffers.
+ */
+class BufferSource : public ReadBufferSource, public WriteBufferSource {
+public:
+  virtual ~BufferSource() {}
+};
+
+/**
+ * Adapter that masquerades a given buffer instance as a ReadBufferSource.
+ */
+class FixedReadBufferSource : public ReadBufferSource {
+public:
+  FixedReadBufferSource(Buffer::Instance& data, bool end_stream)
+      : data_(data), end_stream_(end_stream) {}
+
+  ~FixedReadBufferSource() {}
+
+  StreamBuffer getReadBuffer() override { return {data_, end_stream_}; }
+
+private:
+  Buffer::Instance& data_;
+  const bool end_stream_;
+};
+
+/**
+ * Adapter that masquerades a given buffer instance as a WriteBufferSource.
+ */
+class FixedWriteBufferSource : public WriteBufferSource {
+public:
+  FixedWriteBufferSource(Buffer::Instance& data, bool end_stream)
+      : data_(data), end_stream_(end_stream) {}
+
+  ~FixedWriteBufferSource() {}
+
+  StreamBuffer getWriteBuffer() override { return {data_, end_stream_}; }
+
+private:
+  Buffer::Instance& data_;
+  const bool end_stream_;
+};
+
+/**
+ * Callbacks used by FilterManager to interact with Connection in advanced cases,
+ * i.e. to pause and resume write operation.
+ */
+class FilterManagerCallbacks {
+public:
+  virtual ~FilterManagerCallbacks() {}
+
+  /**
+   * Write data to the connection bypassing filter chain.
+   *
+   * I.e., consider a scenario where iteration over the filter chain is stopped at some point
+   * and then is resumed later via a call to WriteFilterCallbacks::injectDataToFilterChain().
+   *
+   * @param data supplies the data to write to the connection.
+   * @param end_stream supplies whether this is the last byte to write on the connection.
+   */
+  virtual void write(Buffer::Instance& data, bool end_stream) PURE;
+};
+
 /**
  * This is a filter manager for TCP (L4) filters. It is split out for ease of testing.
  */
 class FilterManagerImpl {
 public:
-  FilterManagerImpl(Connection& connection, BufferSource& buffer_source)
-      : connection_(connection), buffer_source_(buffer_source) {}
+  FilterManagerImpl(Connection& connection, BufferSource& buffer_source,
+                    FilterManagerCallbacks& callbacks)
```

Comment:
Instead of all of the callbacks, potentially just add a new `rawWrite(...)` function to the connection interface? Then `write(...)` can be implemented partially using `rawWrite(...)` and I think it would make the code in general simpler to understand? WDYT?

Reply:
`Connection::rawWrite` would give everyone opportunity to bypass filter chain. It's too much of a side-effect for this PR.

What I have right now is a contract between Connection and FilterManager how to resume write operation.

I could rename `FilterManagerCallbacks::write` into `FilterManagerCallbacks::continueWrite` to make my intent more clear.

Reply:
Let me think about this a little more. I'm not crazy about this callback approach, it's kind of confusing.

Reply:
I thought about it a bit and what do you think about creating a `FilterManagerConnection` interface which derives from `Connection`? This interface can add `rawWrite()` and whatever else we need in the future, and then `ConnectionImpl` can just derived from `FilterManagerConnection` but still return `Connection` to all callers? This would allow just continuing to pass `*this` to the filter manager constructor.

I like this a bit better because we don't wind up with multiple sources of calling back into the connection which seems a bit confusing, and we also will save a bit of memory per connection. (I admit this is all a little confusing, but the filter manager code was split out for ease of testing.) WDYT?

Reply:
Yeah, it seems like a simpler solution. I will implement it.

Reply:
Fixed